### ISSUE TITLE
[Feature request] Command rewards

### ISSUE THREAD

**Issue by Sikatsu** [garbagemule/MobArena#348]
Is there a way to give command rewards or give special names and lores to an reward?

**Comment by mibby** [comment]
@garbagemule I'd also like to see command based rewarding as a possible option to receive so you can have a mixture of economy payout, item amount, or a specific command that is ran on leave.

**Comment by garbagemule** [comment]
Hi friends. I'm currently reworking the rewards/items/things API (see the discussion here: https://github.com/garbagemule/MobArena/pull/376), and I would like to know if you guys have any experience with the actual implementation of command-based rewards. Or maybe some examples of plugins that do it in a satisfying way. I assume the commands need to be executed as the "server", rather than impersonating a player and typing a command for them, right? And you would like a means of referencing the player in question as well, right? Can you give me an example of a command you would like to run as a reward? Like `/give garbagemule dirt` could be generified to `/give <player> dirt` or something like that.

**Comment by mibby** [comment]
@slipcor ^ :P 

@garbagemule Running commands such as ```/give garbagemule dirt``` is something that would have a use case, but defining any command to be run from console as the console with a player placeholder would be ideal. Such as if you want a very specific plugin command to run for that user under specific wave conditions - rewarding them after they are no longer part of the arena fight.

I.e. beat wave 100, promote the user via;
```/pex user <user> group set/add [rank]```

Or give a very specific item with NBT tags, something like;
```/minecraft:give <player> minecraft:diamond_axe 1 15 {Unbreakable:1b}

Defining any command to be run from console with the player's name as a playerholder gives infinite reward possibilities for an infinite number of plugins outside of standard item and economy rewarding. You could even have it run a command that globally broadcasts to all players that a user beat a specific wave / boss for example if you allowed for forced rewards on a specific wave rather than a random selection if multiple rewards exist.

**Comment by garbagemule** [comment]
Is the player name the only necessary variable? If so, it would be very simple to implement. As for forcing rewards during the session, that's exactly what upgrade waves were made for. Perhaps that would be enough.

**Comment by mibby** [comment]
Being able to specify any command and run it as console with a player variable, I'd say that would cover everything that I would have a use for.

What I meant by forced rewards is something like if a player beat wave 100, always run the command listed in addition to a random reward type. Something like;

```
    rewards:
      waves:
        every:
          '5': diamond:1, $10, emerald:1
        after:
          '100': $500, 122, 443, 397:5:1
        forced:
          '100': cmd:broadcast <player> just beat the mob arena!
```

Where cmd specifies running the command from console and you could have player: specify running the command from the player. Or console instead of cmd.

**Comment by garbagemule** [comment]
Okay, the functionality in terms of commands is well-defined at this point, so I have something to get crackin' on.

As for forcing rewards, see https://github.com/garbagemule/MobArena/issues/386

**Comment by garbagemule** [comment]
@Sikatsu @mibby I have a working prototype. Will you guys please test it out? :)

This is a build off of the `things` branch with some local changes on top, meaning you get all the refactoring goodness from that branch (basically, you should experience no changes), as well as the ability to run commands as rewards. Here's how it works:

    rewards:
      'after':
        '3': cmd:give <player> flint_and_steel

This will give the player flint and steel on wave 3, but via a command (I couldn't think of anything clever to use, but I'm sure you guys can). Now, this will result in the rewards message being `[MobArena] You just earned a reward: /give <player> flint_and_steel`. I assumed this wouldn't fly, so I added a tedious bit of extra parsing logic, allowing you to give the command a title:

    rewards:
      'after':
        '3': cmd(a wonderful thing):give <player> flint_and_steel

This will result in `[MobArena] You just earned a reward: a wonderful thing`.

---

Okay so, you can give a command a title with a little extra syntax. But is it good? Is it crap? Would you prefer a different syntax? I personally think it looks a bit ugly, but I can't think of a clearer syntax that is also unambiguous to parse (the command itself *has* to be last, and it can't contain commas because MobArena blindly splits on commas).

Suggestions? Feedback? Hit me!

---

Here's the build (for 1.11): https://www.dropbox.com/s/1c5lljfq6pryf8n/MobArena.jar?dl=0

**Comment by garbagemule** [comment]
This has finally been merged to master and released in v0.99 :)
